CodeGen/Generators/QuantityRelationsParser.cs

```diff
@@ -0,0 +1,174 @@
+            quantityDictionary["1"] = pseudoQuantity with { Name = "1" };
+
+            var relations = ParseRelations(rootDir, quantityDictionary);
+
+            // Because multiplication is commutative, we can infer the other operand order.
+            relations.AddRange(relations
+                .Where(r => r.Operator is "*" or "inverse" && r.LeftQuantity != r.RightQuantity)
+                .Select(r => r with
+                {
+                    LeftQuantity = r.RightQuantity,
+                    LeftUnit = r.RightUnit,
+                    RightQuantity = r.LeftQuantity,
+                    RightUnit = r.LeftUnit,
+                })
+                .ToList());
+
+            // We can infer TimeSpan relations from Duration relations.
+            var timeSpanQuantity = pseudoQuantity with { Name = "TimeSpan" };
+            relations.AddRange(relations
+                .Where(r => r.LeftQuantity.Name is "Duration")
+                .Select(r => r with { LeftQuantity = timeSpanQuantity })
+                .ToList());
+            relations.AddRange(relations
+                .Where(r => r.RightQuantity.Name is "Duration")
+                .Select(r => r with { RightQuantity = timeSpanQuantity })
+                .ToList());
+
+            // Sort all relations to keep generated operators in a consistent order.
+            relations.Sort();
+
+            foreach (var quantity in quantities)
+            {
+                var quantityRelations = new List<QuantityRelation>();
+
+                foreach (var relation in relations)
+                {
+                    if (relation.LeftQuantity == quantity)
+                    {
+                        // The left operand of a relation is responsible for generating the operator.
+                        quantityRelations.Add(relation);
+                    }
+                    else if (relation.RightQuantity == quantity && relation.LeftQuantity.Name is "double" or "TimeSpan")
+                    {
+                        // Because we cannot add generated operators to double or TimeSpan, we make the right operand responsible in this case.
+                        quantityRelations.Add(relation);
+                    }
+                }
+
+                quantity.Relations = quantityRelations.ToArray();
+            }
+        }
+
+        private static List<QuantityRelation> ParseRelations(string rootDir, IReadOnlyDictionary<string, Quantity> quantities)
+        {
+            var relationsFileName = Path.Combine(rootDir, "Common/UnitRelations.json");
+
+            try
+            {
+                var text = File.ReadAllText(relationsFileName);
+                var relationStrings = JsonConvert.DeserializeObject<List<string>>(text) ?? [];
+                relationStrings.Sort();
+
+                var parsedRelations = relationStrings.Select(relationString => ParseRelation(relationString, quantities)).ToList();
```

Comment:
Should we throw on duplicates here, with a helpful explanation? 

It probably gives compile errors in the generated code anyway, but it would be helpful if codegen failed early on invalid input to make it easier for contributors to find out what they did wrong.

If we eventually do #1354 , then duplicates could also occur implicitly by one explicit definition conflicting with the implicit definition for division operators.

Another option is to remove duplicates with a HashSet or similar, then fix the document when writing it back. But it may be complicated to know what to remove, in particular with the magic division operators, so it is probably better to just throw and have the author fix their mistake.

Reply:
Good idea, I added a check after all relations are inferred.

Duplicate definitions in `UnitRelations.json` are automatically removed with a `SortedSet`.